### PR TITLE
fqdn: Fix fqdn subsystem correctness issues causing packet drops

### PR DESCRIPTION
**Pull request by odinuge**
tl;dr this fixes a very bad issue in the fqdn subsystem that cause a permanent broken state where certain IPs get stuck in a state where they cannot be used until they are flushed from the fqdn cache and re-added. That is very problematic when the given fqdn keep returning the same IP(s).

This is still work in progress since its a pretty big change, so want to run test to see if its anything very obviously wrong with it.

Implemented together with @sjdot.

This affects v1.15+, and TODO on me to add the correct `Fixes: ` tag.

Comment message
```
    fqdn: Fix fqdn subsystem correctness issues causing packet drops

    To ensure we never leak/fail to upsert identities, we ensure we only add
    IPs that were already present in the local cache.

    The error we are fixing here is the one where the domain is already in
    the FQDN cache, AND a lookup is in flight. If its added to to local
    endpoint cache before the GC ReplaceFromCacheByNames is executed, it
    could result in it being added to the global cache by the GC,
    hence it never being upserted by the dns lookup code. This leads to a
    situation where the toFQDN won't allow egress to the given IP - and the
    only way to recover for the given IP is to wait until its removed from
    the fqdn cache and a new lookup returns the same IP. This is however not
    very likely, since the same domain can continue to return the same IP
    forever, keeping it constantly in the fqdn cache.

    The same partially applies to endpoint created during the GC iteration,
    resulting in the GC evicting some of the IPs upserted to the global
    cache by that endpoint. That situation could also result in a leak where
    its removed from the fqdn cache, but still be leaked in the ipcache.

    This still needs extensive testing to ensure it doesn't uncover other
    issues that rely on the GC upserting new IPs into the global cache - as
    those could now become even more problematic. We have been running a
    version of this in v1.15 and v1.16 with great success, and without this
    patch we see loads of issues and packet drops due to this issue. We have
    not been testing this in high scale in newer versions than those
    mentioned, and due to other changes in the fqdn subsystem, we need to
    sanity check this thoroughly.

    This is still in works, and will be split into multiple commits, and
    potentially multiple PRs.
```

This also has some improvements in terms of performance as well using the existing `UpdateGenerateDNS` benchmark;
```diff
# based on: go test -bench ^BenchmarkUpdateGenerateDNS$ ./ -benchmem -v -benchtime=20x -count=10
                     │   /tmp/old   │            /tmp/new            │
                     │    sec/op    │    sec/op     vs base          │
UpdateGenerateDNS-12   332.2m ± 20%   356.1m ± 13%  ~ (p=0.529 n=10)

                     │   /tmp/old   │               /tmp/new               │
                     │     B/op     │     B/op      vs base                │
UpdateGenerateDNS-12   5.363Mi ± 2%   4.418Mi ± 2%  -17.62% (p=0.000 n=10)

                     │  /tmp/old   │              /tmp/new              │
                     │  allocs/op  │  allocs/op   vs base               │
UpdateGenerateDNS-12   46.10k ± 0%   42.20k ± 0%  -8.46% (p=0.000 n=10)
```

By patching in another case in the benchmark to show FQDNs with a lot of IPs, the diff is even more evident;
```diff
diff --git a/pkg/fqdn/namemanager/bench_test.go b/pkg/fqdn/namemanager/bench_test.go
index dc5a08cef8..8ffb8e20de 100644
--- a/pkg/fqdn/namemanager/bench_test.go
+++ b/pkg/fqdn/namemanager/bench_test.go
@@ -80,6 +80,11 @@ func BenchmarkUpdateGenerateDNS(b *testing.B) {
                        TTL: 60,
                        IPs: []netip.Addr{ip},
                })
+
+               nameManager.UpdateGenerateDNS(context.Background(), t, dns.FQDN("example.com"), &fqdn.DNSIPRecords{
+                       TTL: 60,
+                       IPs: []netip.Addr{ip},
+               })
        }
 }
```

```
# based on: go test -bench ^BenchmarkUpdateGenerateDNS$ ./ -benchmem -v -benchtime=20x -count=10
                     │ /tmp/old-with-change │         /tmp/new-with-change         │
                     │        sec/op        │    sec/op     vs base                │
UpdateGenerateDNS-12          1116.9m ± 66%   350.4m ± 14%  -68.63% (p=0.000 n=10)

                     │ /tmp/old-with-change │         /tmp/new-with-change         │
                     │         B/op         │     B/op      vs base                │
UpdateGenerateDNS-12         471.672Mi ± 0%   5.516Mi ± 3%  -98.83% (p=0.000 n=10)

                     │ /tmp/old-with-change │        /tmp/new-with-change         │
                     │      allocs/op       │  allocs/op   vs base                │
UpdateGenerateDNS-12            59.15k ± 0%   53.20k ± 0%  -10.06% (p=0.000 n=10)
```

Fixes: #issue-number

```release-note
fqdn: Fix fqdn subsystem correctness issues causing packet drops and inconsistent ipcache
```

### CODE REVIEW

File: pkg/fqdn/cache.go [hunk 1]
```diff
@@ -503,6 +549,46 @@ func (c *DNSCache) RemoveKnown(mappings map[netip.Addr][]string) {
...
+
+	entry := &cacheEntry{
+		Name:           name,
+		LookupTime:     lookupTime,
+		ExpirationTime: lookupTime.Add(time.Duration(ttl) * time.Second),
+		TTL:            ttl,
+		IPs:            ips,
+	}
+
+	c.mu.Lock()
```
Comment (odinuge) [inline pkg/fqdn/cache.go]: I'm not too sure this "longer" critical section is still needed after some of the other changes, so will look into that.

File: pkg/fqdn/cache.go [hunk 2]
```diff
@@ -139,6 +140,10 @@ type DNSCache struct {
 	// sent in the Update is lower, the TTL will be overwritten to this value.
 	// Due is only read-only is not protected by the mutex.
 	minTTL int
+
+	// updated is a map tracking the other DNSCaches that have contributed to the given DNSCache
+	// since last GC round
+	updated map[*DNSCache]struct{}
```
Comment (odinuge) [inline pkg/fqdn/cache.go]: Not sold in this structure, but its a decent way of ensuring we don't miss a new endpoint during the GC operation.

File: pkg/fqdn/namemanager/manager.go [hunk 3]
```diff
@@ -5,23 +5,21 @@ package namemanager
 
 import (
 	"context"
+	"fmt"
```
Comment (odinuge) [inline pkg/fqdn/namemanager/manager.go]: Ahh, thought I removed this. I'll remove it so we can run the tests.

**Comment by odinuge** [comment]
This issue can easily be repored by just patching in these two `time.Sleep` calls, and doing some lookups to a fqdn that has a lot of IPs like `test.s3.amazonaws.com`. An example is eg. having a dns and toFQDN policy for `test.s3.amazonaws.com` and running this in a pod; `while true; do timeout 5 wget test.s3.amazonaws.com.; done`.

```diff
diff --git a/pkg/fqdn/messagehandler/message_handler.go b/pkg/fqdn/messagehandler/message_handler.go
index f9e60b6a1e..2d29ec6acc 100644
--- a/pkg/fqdn/messagehandler/message_handler.go
+++ b/pkg/fqdn/messagehandler/message_handler.go
@@ -317,6 +317,7 @@ func (h *dnsMessageHandler) UpdateOnDNSMsg(lookupTime time.Time, ep *endpoint.En
        defer updateCancel()
        updateStart := time.Now()

+       time.Sleep(1 * time.Second)
        dpUpdates := h.nameManager.UpdateGenerateDNS(updateCtx, lookupTime, qname, &fqdn.DNSIPRecords{
                IPs: responseIPs,
                TTL: int(TTL),
diff --git a/pkg/fqdn/namemanager/manager.go b/pkg/fqdn/namemanager/manager.go
index e52edd7b09..d5dea18b16 100644
--- a/pkg/fqdn/namemanager/manager.go
+++ b/pkg/fqdn/namemanager/manager.go
@@ -102,7 +102,7 @@ func New(params ManagerParams) *manager {
                params.JobGroup.Add(job.Timer(
                        dnsGCJobName,
                        n.doGC,
-                       DNSGCJobInterval,
+                       time.Millisecond*500,
                ))

                params.JobGroup.Add(job.OneShot(

```

File: pkg/fqdn/cache.go (hunk 1 above)
Comment (odinuge) [reply pkg/fqdn/cache.go]: After some more thinking, I'm pretty sure the GC can run where the `// GC can run here` comment is placed. That could in theory result in the GC clearing the IPs <-> name mapping from the ipcache. Then, after the `n.cache.Update`, the second `n.cache.Lookup` could return the same IPs as the initial one.


```go
oldCacheIPs := n.cache.Lookup(dnsName)

if n.params.Config.MinTTL > ttl {
	ttl = n.params.Config.MinTTL
}
  // GC can run here
changed := n.cache.Update(lookupTime, dnsName, newIPs, ttl)
if !changed { // Changed may have false positives, but not false negatives
	return false
}

newCacheIPs := n.cache.Lookup(dnsName) // DNSCache returns IPs unsorted
```

It should be possible to reduce this critical section by updating `updateWithEntry` to return true `iff` new IPs are added, and we can remove the whole `ip.SortAddrList` altogether. In production we have seen that code path be pretty expensive.

Comment (odinuge) [reply pkg/fqdn/cache.go]: Updated with a separate commit to resolve ^.

File: pkg/fqdn/cache.go [hunk 4]
```diff
@@ -193,29 +198,31 @@ func (c *DNSCache) Update(lookupTime time.Time, name string, ips []netip.Addr, t
...
-func (c *DNSCache) updateWithEntry(entry *cacheEntry) bool {
-	changed := false
+// This needs a write lock. Returns true if one or more IPs added to the cache
+// were new.
+// It returns two booleans that indicate if the dns cache was updated, and if one
+// or more IPs were new and therefore upserted
+func (c *DNSCache) updateWithEntry(entry *cacheEntry) (bool, bool) {
 	entries, exists := c.forward[entry.Name]
 	if !exists {
-		changed = true
```
Comment (odinuge) [inline pkg/fqdn/cache.go:206]: I figured out this is a bit confusing, so decided to remove. We already know that if we are adding _something_ here, and the fqdn did not exist in the cache - it will always add one or more entries.

**Comment by odinuge** [comment]
cc @christarazi

**Comment by squeed** [comment]
Thanks for the fix! I'll take some time to think it over.

I have a far-out idea: I've _never_ liked the per-endpoint DNS caches. I think they cause more harm than good, as this bug is evidence of. I think they only exist because endpoints have an existing mechanism for writing to disk.

Could you imagine a world where we only have one, global FQDN cache? Persistence could be achieved by using SQLite or even boltdb. Then we could rip this complexity all out and be done with it for good. Plus, we wouldn't have the overhead of writing lots of json just for the endpoint.

**Comment by christarazi** [comment]
I'm also curious about global DNS cache also given that https://github.com/cilium/cilium/pull/42485 is another example of how per endpoint DNS cache has been problematic.

File: pkg/fqdn/cache.go [hunk 5]
```diff
@@ -360,53 +367,88 @@ func (c *DNSCache) GC(now time.Time, zombies *DNSZombieMappings) (affectedNames
...
-// multiple UpdateFromCache invocations. The result is to collect all entries
-// for DNS names in namesToUpdate from each DNSCache in updates, replacing the
-// current entries for each of those names.
+// multiple partialRestoreFromCache invocations. The result is the union of the existing
+// DNSCache and the updates. We do this to ensure that this process does not upsert new
+// entries to the global DNSCache that has not been seen yet. We do this to ensure the
+// dns code path is the one doing the ipcache upsert correctly so the lookups can correctly
+// wait for ipcache propagation of new IPs. We also do this to ensure IPs are correctly upserted,
+// since they will not be upserted during the GC cycle.
 func (c *DNSCache) ReplaceFromCacheByNames(namesToUpdate []string, updates ...*DNSCache) {
```
Comment (odinuge) [inline pkg/fqdn/cache.go]: I also wonder if we should return `oldEntries` from this func and use that for cleaning up stale IP<>name mappings from the ipcache.

I think there in theory is a world where a given IP is upserted after https://github.com/cilium/cilium/blob/6676bac6cf3c3f70da4d27a5853b3ea94055ef7d/pkg/fqdn/namemanager/gc.go#L55

and cleared during the GC cycle. The current implementation would not be able to GC that IP<>name mapping since its not present in `maybeStaleIPs`. Even if its technically not possible today, the guarntee that we can actually catch that case is probably preferable to the current behavior.

**Comment by odinuge** [comment]
Thanks for looking at this @squeed and @christarazi!

I think I overall agree that another rewrite here _could_ be interesting, and it is something we have looked at as well. I think the main reason we want to fix this first is just the fact that the current implementation in v1.15+ is causing a lot of correctness issues in production for us, and we need(ed) to fix that. On our side we will need to backport this, and we are already running a version of this patchset internally to make fqdn policies work.

There is also, given we see that no tests have caught this behavior before, a possibility that a new major rewrite can cause similar style edge-cases - that potentially cause similar types of issues.

I think the per-endpoint cache is good in terms of limiting how much badness a rouge endpoint can cause, although that is not a bullet proof way of avoiding noisy neighbors.

We did also think about other solutions here while developing this patchset - but realized pretty quickly that another solution would require a lot of work, and given all the refactors in the fqdn, identity and policy subsystem from v1.15 to current main - getting that working in a way that can be backported can prove tough. There is also the time and complexity aspect here, since we needed (and still need) to have a function fqdn subsystem to use toFQDN policies now, so fixing the issues in the existing implementation makes sense to us. If we fix the current implementation first and invest time and effort in correctness and performance tests now, it would be much easier and safer to spend time implementing another alternative solution later.

File: pkg/fqdn/messagehandler/message_handler.go [hunk 6]
```diff
@@ -303,8 +303,10 @@ func (h *dnsMessageHandler) UpdateOnDNSMsg(lookupTime time.Time, ep *endpoint.En
 	// consistent if a regeneration happens between the two steps. If an update
 	// doesn't happen in the case, we play it safe and don't purge the zombie
 	// in case of races.
-	if updated := ep.DNSHistory.Update(lookupTime, qname, responseIPs, int(TTL)); updated {
-		ep.DNSZombies.ForceExpireByNameIP(lookupTime, qname, responseIPs...)
+	if updated, upserted := ep.DNSHistory.Update(lookupTime, qname, responseIPs, int(TTL)); updated {
```
Comment (squeed) [inline pkg/fqdn/messagehandler/message_handler.go]: What if we moved writing both caches in to `UpdateGenerateDNS()`? That way we could ensure that the two caches are always in sync. We do need to be careful about lock ordering; always taking the lock of the main cache before that of the endpoint cache. But this makes the updates externally atomic.

**Comment by squeed** [comment]
> On our side we will need to backport this, and we are already running a version of this patchset internally to make fqdn 
policies work.

@odinuge I understand completely. Consolidating caches would be a longer-term effort, separate from this PR.

File: pkg/fqdn/messagehandler/message_handler.go (hunk 6 above)
Comment (odinuge) [reply pkg/fqdn/messagehandler/message_handler.go]: I think we can do so. I did initially think about it but abandoned because of lock ordering (since we have seen a bit too many deadlocks involving this stuff) and the push towards speeding up the dns lookup code. For the former I agree it should be fine if we grab the global lock first and then do the `Update` operation on the local one. And the latter I think should be fine with the speedups in https://github.com/cilium/cilium/pull/42500/commits/4067bc9e4cc531d927a6b3d2023586b8de59452c.

Do you think it makes sense to do that in this PR, or in another one? After some thinking, I still think we want the code to ensure we never upsert new IP<>name mappings during GC, so we won't get rid of that. We also probably want to add some logs when its almost happening to get signal about it, especially if we sync the global and local caches as you are suggesting. I think we also still need to keep a set of the seen local caches inside of the global one as well, to avoid GCing data from new endpoints that were not created in time for the endpoint list call in the beginning of the GC operation. So I partially wonder if keeping the change as-is is best, and then look into this later? What do you think?

File: pkg/fqdn/cache.go [hunk 7]
```diff
@@ -178,7 +183,7 @@ func (c *DNSCache) DisableCleanupTrack() {
 // name is used as is and may be an unqualified name (e.g. myservice.namespace).
 // ips may be an IPv4 or IPv6 IP. Duplicates will be removed.
 // ttl is the DNS TTL for ips and is a seconds value.
-func (c *DNSCache) Update(lookupTime time.Time, name string, ips []netip.Addr, ttl int) bool {
+func (c *DNSCache) Update(lookupTime time.Time, name string, ips []netip.Addr, ttl int, updates ...*DNSCache) (bool, bool) {
```
Comment (odinuge) [inline pkg/fqdn/cache.go]: I think I'll update this to return a struct instead to avoid the `(bool, bool)` and potentially messing up the ordering.

**Comment by christarazi** [comment]
> Thanks for looking at this @squeed and @christarazi!
> 
> I think I overall agree that another rewrite here _could_ be interesting, and it is something we have looked at as well. I think the main reason we want to fix this first is just the fact that the current implementation in v1.15+ is causing a lot of correctness issues in production for us, and we need(ed) to fix that. On our side we will need to backport this, and we are already running a version of this patchset internally to make fqdn policies work.
> 
> There is also, given we see that no tests have caught this behavior before, a possibility that a new major rewrite can cause similar style edge-cases - that potentially cause similar types of issues.
> 
> I think the per-endpoint cache is good in terms of limiting how much badness a rouge endpoint can cause, although that is not a bullet proof way of avoiding noisy neighbors.
> 
> We did also think about other solutions here while developing this patchset - but realized pretty quickly that another solution would require a lot of work, and given all the refactors in the fqdn, identity and policy subsystem from v1.15 to current main - getting that working in a way that can be backported can prove tough. There is also the time and complexity aspect here, since we needed (and still need) to have a function fqdn subsystem to use toFQDN policies now, so fixing the issues in the existing implementation makes sense to us. If we fix the current implementation first and invest time and effort in correctness and performance tests now, it would be much easier and safer to spend time implementing another alternative solution later.

I think we're aligned. I'm in favor of this version of the PR as the fix now to resolve the immediate issue, and then focus on the long-term solution moving forward with infrastructure, tests, and so on.

File: pkg/fqdn/cache.go (hunk 5 above)
Comment (christarazi) [reply pkg/fqdn/cache.go]: This makes sense to me

**Comment by odinuge** [comment]
I've added a new test to catch most edge-cases I'm aware of now. I have some question around how the zombie cache works, but I think it should be fine. The big change here is really that the GC operation will never upsert new ip<>name mappings, to avoid ending in a state where that causes it to not be upserted in the ipcache.

I've also added https://github.com/cilium/cilium/pull/42485 since my tests fail without it. I'll rebase as soon as that one gets merged. 😄 I'll do another check on Monday before marking as ready for review.

**Comment by odinuge** [comment]
While working on this I discovered a set of other fqdn related issues in the ipcache. Its effectively impossible to write proper tests and test this e2e without fixing that issue, so lets work on https://github.com/cilium/cilium/pull/42815 first. 😄

**Comment by odinuge** [comment]
I'll rewrite the tests to use a real ipcache once https://github.com/cilium/cilium/pull/42815 and https://github.com/cilium/cilium/pull/42820 is merged. Without those, tests using the real ipcache all fail due to bugs in the upsert logic.

edit nov 25th: Finally all other deps and the stuff I'm working on have been merged, so I'll start to rebase and update the tests here tomorrow morning.

File: pkg/fqdn/messagehandler/message_handler.go (hunk 6 above)
Comment (squeed) [reply pkg/fqdn/messagehandler/message_handler.go]: > I still think we want the code to ensure we never upsert new IP<>name mappings during GC.

You're right, it's even worse than I realized. Re-reading the GC code with a more careful eye, I don't see how we can ever safely GC without doing a stop-the-world lock.

The basic gist of the GC method is:

1. get all in-use per-endpoint caches
2. For every endpoint, remove expired names
3. For every changed name, force-set the global cache to the "sum" of the per-endpoint caches

But what if a new endpoint comes in to existence between steps 1 and 3? Any names we resolve simply won't be there. As a short-term fix, I think we have to take the main DNSCache lock during GC. And that does, IHMO, warrant being in this PR. Thoughts?

Comment (odinuge) [reply pkg/fqdn/messagehandler/message_handler.go]: > You're right, it's even worse than I realized. Re-reading the GC code with a more careful eye, I don't see how we can ever safely GC without doing a stop-the-world lock.

Did you read the version in this PR, or the version in `main`? But yeah, I agree the current `main` state is pretty bad. We are unable to use v1.15 without a version of this patchset.

> But what if a new endpoint comes in to existence between steps 1 and 3? Any names we resolve simply won't be there. As a short-term fix, I think we have to take the main DNSCache lock during GC. And that does, IHMO, warrant being in this PR. Thoughts?

This is exactly the reason we have proposed this change tbh. There are multiple race conditions, and all of them have their own issues.

Can you read through this test https://github.com/cilium/cilium/pull/42500/commits/9417a74b85f21adb259b0746f7129669b3aa8e1b#diff-adb29c55afd0ee7e6ec4eba2bac390e7967c33c91b6d3d252a1d2445c712e95aR223-R246? Also this part of the code is relevant for that use case https://github.com/cilium/cilium/pull/42500/commits/c462271e782e6034c5fb9699c86b6cca86a3b4ad#diff-7417ec0b53fdf1f35c5cca08a65475512c12491dede47711cf891bfef6b6a2e9R430-R436?

Does that answer the question above, or am I missing something?

There might be more races here, especially around restore logic, given I don't really understand all the `cell` dependencies, as well as them changing rapidly. However there is no way iirc, with the proposed PR, an endpoint can be created during the GC cycle and have its lookups cleared from the global cache. Since we in this patchset ensure we always "store" all the per-endpoint caches used to update the global. Then during GC we also ensure we use those when clearing and re-populating the cache.

Maybe we should also deal with zombies as well in that setup/upsert, but I'm not too sure if thats required tbh. We could write a test for it tho, and then see if that works.

Comment (odinuge) [reply pkg/fqdn/messagehandler/message_handler.go]: > As a short-term fix, I think we have to take the main DNSCache lock during GC. And that does, IHMO, warrant being in this PR. Thoughts?

We can probably do that as well here, but I'm worried about lock contention, slow DNS lookups and potential deadlocks tbh. So that would require a lot more investigation.

Comment (odinuge) [reply pkg/fqdn/messagehandler/message_handler.go]: And I think in the end, the most important thing is that we _never_ do something that results in a state like today, where something is present in the global fqdn cache but missing in the ipcache. That failure mode results in a permanently broken state where new lookups return the same IP(s) still does not fix the issue.

A race that somehow removes it from the global cache (and then maybe the ipcache) can easily recover by a similar lookup to the same IP. Such a race is _not great_ for sure, and we want to avoid it - but its not catastrophic in the same way.

Comment (squeed) [reply pkg/fqdn/messagehandler/message_handler.go]: Sorry for the noise, I overlooked the functionality of `c.updated` from the diff.

**Review by squeed — APPROVED** [review]

**Comment by bimmlerd** [comment]
Removing myself - my review is covered and I'm lacking cycles at the moment to give this enough time to understand what's going on. Thanks!

**Review by christarazi — APPROVED** [review]
Do you plan to add more testing now that the previous PRs have been merged? Otherwise, LGTM

File: pkg/fqdn/messagehandler/message_handler.go (hunk 6 above)
Comment (odinuge) [reply pkg/fqdn/messagehandler/message_handler.go]: > @squeed: A close reading of the code makes me suspect we may also never clean up names held by endpoints deleted between GC runs. I can look in to this in a separate PR.

We already tackled that in https://github.com/cilium/cilium/pull/42485 iirc. That patch is already in this patchset since the tests depend on it, ref. https://github.com/cilium/cilium/pull/42500#issuecomment-3532713594. So I'm pretty sure it should be fine now. I'll rebase and remove it now that its merged 😄 


> @odinuge: Maybe we should also deal with zombies as well in that setup/upsert, but I'm not too sure if thats required tbh. We could write a test for it tho, and then see if that works.

Did check in on this, and we only do evict entries like this in `func (c *DNSCache) GC` atm. And that only runs if the endpoint was created before the `func (n *manager) doGC` dumps endpoints. So I think the current tracking of `c.updated` works as expected on these edge cases.

**Comment by odinuge** [comment]
Sweet, thanks for reviewing! Its been a bit of a chaos with all the other related PRs, so I'll rebase when I get time today or tomorrow and remove the cherry-pick of https://github.com/cilium/cilium/pull/42485 now that its merged - and I'll rewrite tests to use the proper ipcache now that https://github.com/cilium/cilium/pull/42815 and https://github.com/cilium/cilium/pull/42820 are merged. Hopefully I can get that done before the weekend.

**Comment by squeed** [comment]
@odinuge no stress! I've marked this as "needs-rebase" so it doesn't accidentally get merged. I think you can remove the label yourself, or ping one of us to do so.

**Comment by odinuge** [comment]
I've rebased and updated the tests now - and they all seems to pass now!
